--- a/clay.h    2026-01-22 14:54:04.382314498 -0800
+++ b/clay.h    2026-01-22 14:54:53.241299528 -0800
@@ -28,6 +28,7 @@
 #ifndef CLAY_HEADER
 #define CLAY_HEADER
 
+/*
 #if !( \
     (defined(__cplusplus) && __cplusplus >= 202002L) || \
     (defined(__STDC__) && __STDC__ == 1 && defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) || \
@@ -36,6 +37,7 @@
 )
 #error "Clay requires C99, C++20, or MSVC"
 #endif
+*/
 
 #ifdef CLAY_WASM
 #define CLAY_WASM_EXPORT(name) __attribute__((export_name(name)))
