diff --git a/source/graphics.cpp b/source/graphics.cpp
index c5eb3be..8d997ae 100644
--- a/source/graphics.cpp
+++ b/source/graphics.cpp
@@ -484,7 +484,7 @@ FontFaceCache::FontFaceCache()
 
 FontFaceCache* fontFaceCache()
 {
-    thread_local FontFaceCache cache;
+    static FontFaceCache cache;
     return &cache;
 }
 
diff --git a/source/lunasvg.cpp b/source/lunasvg.cpp
index 2cb72d1..aea64e3 100644
--- a/source/lunasvg.cpp
+++ b/source/lunasvg.cpp
@@ -462,8 +462,8 @@ Box Document::boundingBox() const
 
 void Document::updateLayout()
 {
-    SVGLayoutState state;
-    m_rootElement->layout(state);
+    auto state = std::make_unique<SVGLayoutState>();
+    m_rootElement->layout(*state);
 }
 
 void Document::render(Bitmap& bitmap, const Matrix& matrix) const
diff --git a/source/svgelement.cpp b/source/svgelement.cpp
index dee7ca2..5224991 100644
--- a/source/svgelement.cpp
+++ b/source/svgelement.cpp
@@ -441,9 +441,9 @@ void SVGElement::layoutChildren(SVGLayoutState& state)
 
 void SVGElement::layout(SVGLayoutState& state)
 {
-    SVGLayoutState newState(state, this);
-    layoutElement(newState);
-    layoutChildren(newState);
+    auto newState = std::make_unique<SVGLayoutState>(state, this);
+    layoutElement(*newState);
+    layoutChildren(*newState);
 }
 
 void SVGElement::renderChildren(SVGRenderState& state) const
diff --git a/source/svglayoutstate.cpp b/source/svglayoutstate.cpp
index e86cfab..91ce122 100644
--- a/source/svglayoutstate.cpp
+++ b/source/svglayoutstate.cpp
@@ -13,7 +13,8 @@ static std::optional<Color> parseColorValue(std::string_view& input, const SVGLa
     }
 
     plutovg_color_t color;
-    int length = plutovg_color_parse(&color, input.data(), input.length());
+    std::string str(input);
+    int length = plutovg_color_parse(&color, str.c_str(), str.length());
     if(length == 0)
         return std::nullopt;
     input.remove_prefix(length);
